### ISSUE TITLE
Performance suggestions in Wiki?

### ISSUE THREAD

**Issue by BHare1985** [mapnik/mapnik#2890]
I am looking for performance suggestions, looked at wiki and ctrl+f and searched for the word performance, resulted in nothing.

Assume you don't have any notes/suggestions on this.

A few questions I was wondering:

Is it better to store data in WKT, WKB, GeoJSON in terms of performance with mapnik alone. I remember seeing that mapnik tries to do some R-tree stuff itself. I understand that depending on datasource they will handle it differently. (PostGIS will do R-trees with WKB, MSSQL will do a B-tree thing with WKB, and MongoDB will use Geohash with GeoJSON)

What does Mapnik prefer?

Also, should I store the geometry/geography in the same SRID as I am projecting it on the map. Right now I store it in 4326 and project to 3857 (as I seen many examples do), but it would seem if I store it in 3857 then its less work mapnik has to do?

**Comment by flippmoke** [comment]
Projection only  matters if you have to reproject, what projection are you using for your map. PostGIS obviously has network latency, but could be better if the dataset is large. GeoJSON, WKB, and WKT will take probably more time for the IO then for performance of the datasource itself. Out of the 3 databases listed, I would use PostGIS for most operations, but not saying the others are bad choices.

**Comment by flippmoke** [comment]
Also are you trying to render images or vector tiles?

**Comment by BHare1985** [comment]
I am using a projection onto google maps (3857). I am rendering images.

Right now I am using MSSQL with a plugin: https://github.com/gaspardle/mapnik-mssql

But, we are looking into options for Redshift (Doesnt support Spatial indices/PostGIS) and MongoDB. If we were to go with Redshift I was thinking of ways I could effectively use mapnik as the spatial filter, maybe pass it WKB as a memory datasource but it seems too inefficient in some cases would need load GBs of data to the memory datasource.

**Comment by flippmoke** [comment]
Yes, leaving it in 3857 will make it slightly faster, especially if your queries are pulling in a ton of data. You will probably improve performance more by working on indexing in your database or time optimizing queries. 

What sort of data are you using? Point, line, vector, raster? What are you trying to make?

**Comment by BHare1985** [comment]
Right now we are using mapnik to do just lines, specifically every road in the US, and using grainstore to dynamically style the road segments based on traffic data. In the future we will be applying mapnik to do a mix of geometry but mostly points.

**Comment by flippmoke** [comment]
Do you have a good caching layer? Whats your stack? Custom framework I assume?

**Comment by BHare1985** [comment]
We won't be caching as just about every tile request will be unique, based on search parameters. We are considering tile caches later on, but probably only for higher zooms (ones that take way too long to load) I have forked from an old Windshaft commit and have got it working with windows/MSSQL: https://github.com/BHare1985/Windshaft . This uses grainstore and redis for doing dynamic cartoCSS to Mapnik XML encoding using a backend of tilemill (which is having issues with windows right now). Working on hiding our SQL statements from the end user, or at least not in every tile GET request.

We are evaluating options as far as database/datasources go, because while we are getting usable performance with MSSQL and the 20 million segments we have now for geometry (mapnik does a good job of rendering at zoom 10-20), we will need to join the geometry with our traffic data which is 2 billion records, growing to 200 billion. This performance need is on our datasource, and not so much mapnik as so far we have been pleased with its performance.

**Comment by flippmoke** [comment]
I just have experience with these stacks so I like to help when I can. I am assuming this is temporal data? Its going to be really important for your tree structure that you probably think about breaking out those points by time somehow. [Here is something I have written on this topic before](https://www.reddit.com/r/gis/comments/3837wa/need_suggestions_for_optimization_strategy_when/crslp8y). 

I bet you would be surprised if you tried that your requests were probably quite often not that unique. Another option for speed as the dataset grows would be to simply cut vector tiles where you know there will be data, and update those on demand. Then you could "overzoom" and render a several levels higher (deeper). We actually do that quite a lot in production at Mapbox. 

I would look at some of the stuff we do in node currently:

* https://github.com/mapbox/tilelive-bridge
* https://github.com/mapbox/tilelive-vector

If you haven't heard Dane's talk on vector tiles too, I would encourage that as well

* https://www.youtube.com/watch?v=se2cd3BMYRY

If you are using a windows environment, I would love it if you had any problems that you passed on your issues to @BergWerkGIS :+1:

**Comment by BHare1985** [comment]
Thanks for the resources, I am new to GIS/mapping so any help is appreciated.
I watched the video and actually think they are an attractive solution, I didn't like the idea of doing 2 requests one for the image and one for the grid.

The idea of separating time and points by a bounding box seems like a great idea, but a bounding box with standard indices isn't a solution for line when the it can span multiple tiles. Perhaps if storage wasn't an issue one could store every vertex to a polyline in a table and then use that as a basis to determine which lines will need to be selected for that tile but I am not sure if it will be worth it, as the table will be x the size of the original table where x is the average vertices in a linestring.

Ideally, I'd love to cache the actual geometry of the lines in a vtile and then just color the lines accordingly based on data I get in real time. I am not sure if this is possible with vtiles, I heard something about adding vtiles ontop of other vtiles. Would it be possible to cache all the geometry and features to a vector tile, and then on demand get the geom vtile from disk and add the time data to it and re-style using mapnik xml/cartocss?

**Comment by flippmoke** [comment]
Can you have a background process that simply makes the vector tiles based on changing conditions?

**Comment by BHare1985** [comment]
Sure, that's an option but I am not sure how it would be done. The issue I didn't think about last night was even if I was to cache the geometries in vtiles, I would still need to do a spatial lookup to know which time data to get for those geometries.

**Comment by flippmoke** [comment]
I would assume as your data is updated, you send a message on a queue to update a tile in your store. If tiles have a limited lifespan you could just tag the tile's time when it is created. If you return an out of date tile and there is no data update, it could be stale and nothing is to be rendered. This is quick though because you know exactly what row in the database to get each time. You could also use S3 or a NoSQL database as well to store the tiles if you wanted.

**Comment by BHare1985** [comment]
I am not sure I follow you much.  Lets say I have a tiles that are requested of the city of detroit. At any given moment there could be 2-3 simultaneously users viewing different styles of the same geometry. One user could be looking at the traffic data of the roads for a certain day of the week for the past 3 months, another could be looking at the average for all days in the past 2 years.

Without vtiles, the way I would do it now is get the geometry from the database and JOIN it with the data I need to style it. For each user I have to select the same geometry over and over, when really all I need to do is select the data that styles the geometry.

I haven't found any examples of even getting vtiles running and rendering with mapnik, only things to go on our the tests in each repo.

My hope was to render every vtile with geometry for tiles I will need (GB to TB of data), then upon a new map request it will query a database/noSQL for the data associated for that specific query for that tile location and return a vtile, or a style. I then take the original vtile with geometry and style it based on the data from the database.

Is this possible?  

Do you have any examples of completed and working examples using vtiles and rendering images via mapnik? Similar to how some mapbox repos come with an examples folder that uses merc.shp files and a simple leaflet html page for showing it working.

**Comment by danpat** [comment]
BHare1985 - how many users do you anticipate?  I've found re-rendering the entire geometry, joined with style metadata has been fairly performant as long as you're on the right side of the load curve.  I'm dynamically colouring roads/trails based on an external metadata feed, fairly similar to your description.  I generate tiles on the fly and cache for 5 minutes, my rendering depends not just on the metadata, but the time-of-access as well.  I don't have your problem of "multiple temporal views for the same tile" however.

One bit finding is that the vast majority of the map never gets visited, I found it very wasteful to try to pre-render everything frequently.  Good indexing and dynamic rendering with some caching are working well for me.

If you want to go down the path of pre-rendering vtiles, then just style them on demand, this is what you'll need to plug in:

https://github.com/mapbox/mapnik-vector-tile/blob/master/src/vector_tile_datasource.hpp.  

Use that instead of the postgis datasource, and generate the vtiles on disk somewhere else.

Optionally, you could also look at `mapbox-gl` and consider styling the tiles on the client side, saving you a lot of load.  This sounds like an ideal use case for trying to shift the load client-side.  This will, of course, depend on what your clients support.
